app/src/main/java/net/gsantner/markor/frontend/textview/HighlightingEditor.java

```diff
@@ -546,4 +462,182 @@ public int setSelectionExpandWholeLines() {
+                        while (matcher.find()) {
+                            _maxNumber--;
+                        }
+                    }
+                }
+
+                @Override
+                public void onTextChanged(CharSequence s, int start, int before, int count) {
+                    if (before == 0 && count > 0) {
+                        CharSequence added = s.subSequence(start, start + count);
+                        matcher = pattern.matcher(added);
+                        while (matcher.find()) {
+                            _maxNumber++;
+                        }
+                    }
+                }
+            });
+        }
+
+        public void setTextSize(float textSize) {
+            if (_lastTextSize != _paint.getTextSize()) {
+                _lastTextSize = _paint.getTextSize();
+            }
+            _paint.setTextSize(textSize);
+        }
+
+        public int getNumberDigits(int number) {
+            if (_lastMaxNumberDigits != _maxNumberDigits) {
+                _lastMaxNumberDigits = _maxNumberDigits;
+            }
+
+            if (number < 10) {
+                _maxNumberDigits = 1;
+            } else if (number < 100) {
+                _maxNumberDigits = 2;
+            } else if (number < 1000) {
+                _maxNumberDigits = 3;
+            } else if (number < 10000) {
+                _maxNumberDigits = 4;
+            } else {
+                _maxNumberDigits = 5;
+            }
+            return _maxNumberDigits;
+        }
+
+        public boolean isOutOfLineNumbersArea() {
+            if (_visibleArea.top > _lineNumbersArea.top && _visibleArea.bottom < _lineNumbersArea.bottom) {
+                return false;
+            } else {
+                _lineNumbersArea.top = _visibleArea.top - _visibleArea.height();
+                _lineNumbersArea.bottom = _visibleArea.bottom + _visibleArea.height();
+                return true;
+            }
+        }
+
+        /**
+         * Draw line numbers.
+         *
+         * @param canvas the canvas on which the line numbers will be drawn.
+         */
+        public void draw(final Canvas canvas) {
+            if (!_editor.getLocalVisibleRect(_visibleArea)) {
+                return;
+            }
+
+            final CharSequence text = _editor.getText();
+            final Layout layout = _editor.getLayout();
+            if (text == null || layout == null) {
+                return;
+            }
+
+            final int offsetY = _editor.getPaddingTop();
+            final int layoutLineCount = layout.getLineCount();
+
+            // Only if the visible area is out of current line numbers area or the layout line count changed
+            // We make a single pass through the text to determine
+            // 1. y positions and line numbers we want to draw
+            // 2. max line number (to gauge gutter width)
+            if (isOutOfLineNumbersArea() || layoutLineCount != _lastLayoutLineCount) {
```

Comment:
I don't think we should test this. I think we should redraw line numbers for _every_ `onDraw`.

This will automatically account for font size changes, highlighting state changes etc etc.

The performance penalty seems small.

Reply:
Also, not doing this test will eliminate the need for many state tracking variables, which is always good.

Reply:
Note this will still redraw line numbers for every onDraw via using value of `List` assigned last time, and only recalculate it in the loop when the status changed, instead of executing the loop every onDraw.

Reply:
As mentioned with other comments, I don't see a visible performance penalty from just running this loop every time. IMO this is kind of a premature optimization.